.github/workflows/ci.yaml

```
@@ -77,12 +77,61 @@ jobs:
           node-version-file: '.nvmrc'
           cache: yarn
       - run: corepack enable yarn
-      - run: SKIP_BUILD_INTERNAL_DEPENDENCIES=true yarn install
-      - run: yarn build:internal-dependencies
+      - name: Install dependencies
+        run: yarn workspaces focus tupaia @tupaia/database @tupaia/types
+      - run: yarn run build:from '{database,types}'
       - run: ./packages/devops/scripts/ci/validateTypesAndDbSchemaInSync.sh
 
+  validate_lockfile:
+    name: Validate lockfile
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version-file: '.nvmrc'
+          cache: yarn
+      - run: corepack enable yarn
+      - run: yarn install --immutable
+
+  package_list:
+    name: Generate package list
+    runs-on: ubuntu-latest
+    outputs:
+      PACKAGES: ${{ steps.set-matrix.outputs.PACKAGES }}
+    steps:
+      - uses: actions/checkout@v4
+      - id: set-matrix
+        run: |
+          echo PACKAGES=$(
+            yarn workspaces list --json |
+              jq --compact-output --raw-output --slurp \
+                '[.[] | .name] | map(select(. != "tupaia" and . != "@tupaia/devops" and . != "@tupaia/e2e"))'
+          ) >>"$GITHUB_OUTPUT"
+
+  build:
+    name: Build
+    needs: package_list
```

Do we also need this?
```suggestion
needs: [validate_lockfile, package_list]
```

---

sounds good!